pyquil/tests/test_paulis.py

```diff
@@ -421,15 +421,21 @@ def test_exponentiate_identity():
     generator = PauliTerm("I", 1, 1.0)
     para_prog = exponential_map(generator)
     prog = para_prog(1)
-    result_prog = Program().inst([X(0), PHASE(-1.0, 0), X(0), PHASE(-1.0, 0)])
+    result_prog = Program()
     assert prog == result_prog
 
     generator = PauliTerm("I", 10, 0.08)
     para_prog = exponential_map(generator)
     prog = para_prog(1)
-    result_prog = Program().inst([X(0), PHASE(-0.08, 0), X(0), PHASE(-0.08, 0)])
+    result_prog = Program()
     assert prog == result_prog
 
+    pop = 2.1 * sI(0)
+    expprog = exponential_map(pop)
+    assert expprog(0.5) == Program()
+    pop.__dict__["_ops"][0] = "Y"
```

Comment:
This feels a little fragile.  Would it be reasonable to replace this with a similar test that instead exercises the `is_zero` case? Then you'd only have to modify `pop.coefficient` here no? (which seems preferable since `coefficient` is at least a "public" attribute). I guess that would no longer make this an identity test, but it could be moved into a new test case or even `test_exponentiate_prog` above.

Reply:
Just to be super explicit: It looks like you see that I want a test that would have failed before this PR, and you think this is a reasonable want. I like the idea of using `is_zero`. I'm not super familiar with how APIs are determined in python. But, you don't need a degree in pythonology to see that `pop.__dict__["_ops"][0]` is not in the API. I guess a data field (attribute) like `coefficient` is in by default ? It's certainly more API'ish.

What I most don't like about this test is that the rest of the file tests the API. Testing implementation that is not part of the API should, if it's desirable at all, go elsewhere. The point of the test is anticipating someone doing something they shouldn't be doing anyway.

Reply:
The convention in python is that any attribute prefixed with one (or more) underscores is considered a "private" implementation detail, while attributes without a leading `_` are part of the "public" API. (Double underscores on either side are a special case that are reserved for Python).

In general I'm not opposed to adding tests that document an implementation detail, but I'd also be fine in this case if we just dropped the test or else switched to testing the `is_zero` case since (I think) we could then stick to accessing the "public" API.

Reply:
Regarding underscores. Just a question, don't some people go further and demand the use of accessors for public attributes ?

It's not clear to me why `coefficient` has been made public and the operators private. My instinct is that `PauliTerm`s should be immutable (if only by explicit convention). But, that's a another issue. And it's much more likely that people have been mutating `coefficient`.

Reply:
Using explicit accessors like `get_foo` is rare in python, mostly because Python has a concept of [Descriptors](https://docs.python.org/3/howto/descriptor.html) and  ["properties"](https://docs.python.org/3/howto/descriptor.html#properties) that let you transparently wrap attribute lookup and assignment in a call to a function getter / setter.

Conventions certainly differ, but generally speaking any attribute that doesn't start with a leading underscore is considered "public". As far as I know, we follow this convention in pyquil.

As to why the `coefficient` attribute was made "public", I couldn't say.  It's possible that it was just an oversight and never intended to be mutated, but given that it's already "public", we're stuck with it so to speak.

Reply:
I somehow forgot that `pop.__dict__["_ops"][0]` is not necessary.  `pop._ops[0]` will do. (Hmm `pop` is not a good variable name, maybe `term` for consistency with the other tests, or `pauli_term`).  Still `_ops` has one underscore and `coefficient` does not. However, reviewing the class `PauliTerm` it seems very likely that `PauliTerm` objects are meant to be immutable. Unfortunately, this is not stated explicitly. I think the underscore is there because the class provides interfaces the operators that are no-so tied to the implementation, whereas `coefficient` is straightforward. Another piece of evidence: The class function `term_with_coeff` says it is to "Change the coefficient of a PauliTerm". But, what it actually does is return a copy with a different coefficient.

The issue of mutability is directly relevant here, because it determines whether we are breaking the API, even a little bit. What to do ? I think continuing with this PR and making a note in the changelog, as you suggest is the best course. But, we should come to a consensus on whether `PauliTerm` is mutable. WDYT ?

Reply:
> Hmm pop is not a good variable name, maybe term for consistency with the other tests

I almost made this comment, but didn't want to risk NITPICK OVERLOAD :)

> The issue of mutability is directly relevant here, because it determines whether we are breaking the API, even a little bit. What to do ? I think continuing with this PR and making a note in the changelog, as you suggest is the best course. But, we should come to a consensus on whether PauliTerm is mutable. WDYT ?

I think the fact that the `coefficient` attribute is "public" means that `PauliTerm` is technically mutable, regardless of what the original intention was or whether we like it now. We could change the API to enforce immutability (or at least more strongly discourage mutability), but that seems like a bigger breaking change and something that should only be done in a major version bump.

Reply:
nitpick away!

I meant to ask (but... verbosity overload) how closely we follow semver. If we are strict, then neither of the two changes may be made until a major version bump. It also means we are committed to mutability. (You want spooky action at a distance ? I'll show you spooky action at a distance!) I suppose we could state in the doc for each new interface function that this function only works as advertised if the input data is not subsequently mutated, and we stick to it internally. But, what about composing with their previous code, or other external code ?

The percentage of users who care to know what semver is and who have looked at `PauliTerm` and know or care that `coefficient` may be changed, but not the ops is likely zero. I mean, I don't think we should rely on the presence or absence of an underscore to inform developers or users, at least in this case, where the letter and the intent do not coincide. Making clear what we assume may and may not be done to `PauliTerm` would cut down on the length of discussions like this.

Reply:
> I meant to ask (but... verbosity overload) how closely we follow semver.

I'm not sure how closely we followed it in the past. I think it's up to us now!

> I don't think we should rely on the presence or absence of an underscore to inform developers or users, at least in this case, where the letter and the intent do not coincide. Making clear what we assume may and may not be done to PauliTerm would cut down on the length of discussions like this.

I hear what you're saying, and there is no substitute for good documentation, but the leading underscore thing is a widely-followed Python convention, enshrined in the "official" style guide, PEP8

https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces

And since the current docs don't explicitly say "do not mutate", someone might reasonably assume that mutating `coefficient` is fair game. How likely is it? 🤷‍♂